src/Data/Monoid/Inf.hs

```diff
@@ -75,6 +76,51 @@ instance Ord a => Monoid (Inf Neg a) where
   mempty = Infinity
   mappend = (<>)
 
+instance Applicative (Inf p) where
+    pure = Finite
+    Infinity <*> _ = Infinity
+    _ <*> Infinity = Infinity
+    Finite f <*> Finite x = Finite $ f x
+
+instance Monad (Inf p) where
+    Infinity >>= _ = Infinity
+    Finite x >>= f = f x
+
+instance Bounded a => Bounded (NegInf a) where
+    minBound = Infinity
+    maxBound = Finite maxBound
+
+instance Bounded a => Bounded (PosInf a) where
+    minBound = Finite minBound
+    maxBound = Infinity
+
+instance Num a => Num (Inf p a) where
```

Comment:
After thinking about it more, I don't think these `Num`, `Fractional`, and `Floating` instances make sense, for multiple reasons.  For example, `negate` leafs `Infinity` alone but it should really switch negative and positive infinity (which is not possible given this type).  This means multiplication does not work either, since e.g. negative infinity times itself should be positive infinity.  `recip` should arguably send `Infinity` to `Finite 0` and vice versa.  And some of the `Floating` methods don't make sense either, e.g. `sin Infinity = Infinity`: `sin` does not have a limit as its argument goes to infinity, but it certainly doesn't yield `Infinity`.